lib/cache/cache.go

```diff
@@ -1294,43 +1285,73 @@ func (c *Cache) GetNodes(ctx context.Context, namespace string, opts ...services
 	return rg.presence.GetNodes(ctx, namespace, opts...)
 }
 
+// getNodesWithFnCache implements TTL-based caching for the GetNodes endpoint.  All nodes that will be returned from the caching layer
+// must be cloned to avoid concurrent modification.
+func (c *Cache) getNodesWithFnCache(ctx context.Context, rg readGuard, namespace string, opts ...services.MarshalOption) ([]types.Server, error) {
+	ta := func(_ []types.Server) {} // compile-time type assertion
```

Comment:
I realize that this looks pretty out of place, but its something I feel pretty strongly is the correct way to do this.  Here's why:

Using the closure as a type assertion lets us write the type once, and assert it both inside and outside the closure, essentially guaranteeing that the closure really returns what our type assertions says it does.  If it doesn't, it won't compile. For a function like this (which is robustly tested), thats super unnecessary, and a runtime assertion would be fine.  This isn't about this function though.  Its about a consistent pattern that prevents footguns.  A bunch of methods use the same exact pattern.  Not all of our methods have equal test coverage, and I've seen bad runtime type assertions get missed by tests on a number of occasions.  This particular pattern comes up so frequently, that any refactoring is going to be a tedious process with lots of find and replace.  Any refactoring of these common resource getter methods tends to be tedious and repetitive anyhow, since there are a bunch of nearly identical methods spread throughout the codebase.  Basically, mistakes are gonna happen.  I decided to use a standardized pattern of type assertions like this when I first added the ttl-cache, because it seemed like a simple and effective way to prevent dummies like me from causing themselves a huge headache down the line.

Reply:
I think, in practice, this is giving us very few guarantees.

The assertion on line 1296 / fnCache.Get guards us against a change of typing in GetNodes - I'd argue that a breakage there is unlikely / easy to spot, but it does have some value. (It doesn't have to be in the form of a func assert, though.)

The assertion on line 1303, after the type cast from interface{} to []types.Server, isn't doing anything - `cachedNodes` is guaranteed to be `[]types.Server` by the cast. We should be checking the Go assertion result there instead.

Dropping it would be the preferred outcome for me, but I won't hold you on this.

Reply:
Ah, I think I'm not explaining myself well.  The purpose isn't to check the output of `GetNodes`, or to check that the variable we return from the outer function is what it says it is (the return type of the outer function does this).

The purpose of `ta` is to guarantee that the value returned by the inner function (which is cast to an `interface{}`) is identical to the type that the outer function expects (where the same value is cast back from an `interface{}` to  a concrete type).

Because `ta` is called on both the return value of the inner closure, and the post-conversion value of the outer closure, we guarantee that the inner and outer closures are using the same type.  Here is a simplified example of the same pattern:

```golang
func Outer() T {
    ta := func(_ T) {}
    iface := func() interface{} {
        innerValue := getInnerValue()
        ta(innerValue) // innerValue must be type T for this to compile
        return innerValue
    }()
    outerValue := iface.(T)
    ta(outerValue)  // outerValue must be type T for this to compile
    return outerValue
}
```

If we relied on normal golang type assertions, a conflict between the return type of the inner closure, and the expected type in the outer closure would not be detectable until runtime (which brings us back to the aforementioned issues w/ imperfect test coverage, and refactoring being a huge pain).

Reply:
It _just_ took me a bunch of Slack chats and an in-person sync, but I finally got it. Thanks for the patience.

Like I mentioned, let's check the type cast, but otherwise LGTM.